File: 648. Replace Words.java
class Solution {
    public String replaceWords(List<String> dictionary, String sentence) {
        Trie db = new Trie();
        for (String word : dictionary) db.add(word);
        List<String> answer = new ArrayList<>();
        for (String word : sentence.split(" ")) {
            answer.add(db.find(word));
        }
        return String.join(" ", answer);
    }
}

class Trie {
    Trie[] children;
    boolean isWord;

    Trie() {
        this.children = new Trie[26];
    }
    
    void add(String word) {
        Trie current = this;
        for (int i = 0; i < word.length(); i++) {
            int c = word.charAt(i) - 'a';
            if (current.children[c] == null) {
                current.children[c] = new Trie();
            }
            current = current.children[c];
        }
        current.isWord = true;
    }

    String find(String word) {
        StringBuilder sb = new StringBuilder();
        Trie current = this;
        for (int i = 0; i < word.length(); i++) {
            int c = word.charAt(i)-'a';
            if (current.isWord) return sb.toString();
            if (current.children[c] == null) return word;
            sb.append(Character.toString('a' + c));
            current = current.children[c];
        }
        return sb.toString();
    }
}

File: 648. Replace Words.go
func replaceWords(dictionary []string, sentence string) string {
    db := newTrie()
    for _, word := range dictionary {
        db.add(word)
    }
    answer := []string{}
    for _, word := range strings.Split(sentence, " ") {
        answer = append(answer, db.find(word))
    }
    return strings.Join(answer, " ")
}

type Trie struct {
    children []*Trie
    isWord bool
}

func newTrie() *Trie {
    return &Trie {
        children: make([]*Trie, 26, 26),
        isWord: false,
    }
}

func (t *Trie) add(word string) {
    for _, c := range(word) {
        if t.children[c-'a'] == nil {
            t.children[c-'a'] = newTrie()
        }
        t = t.children[c-'a']
    }
    t.isWord = true
}

func (t *Trie) find(word string) string {
    answer := ""
    for _, c := range word {
        if t.isWord {
            return answer
        }
        if t.children[c-'a'] == nil {
            return word
        }
        answer += string(c)
        t = t.children[c-'a']
    }
    return answer
}
